src/Parsers/tests/gtest_format_hiliting.cpp

```diff
@@ -0,0 +1,303 @@
+            return true;
+
+        if (left_it == left.end() || right_it == right.end())
+            return false;
+
+        // Lookup one character.
+        // Check characters match.
+        // Redundant check, given the hilite-ignorant comparison at the beginning, but let's keep it just in case.
+        if (*left_it != *right_it)
+            return false;
+
+        // Check hilites match if it's not a whitespace.
+        if (!std::isspace(*left_it) && left_hilite != right_hilite)
+            return false;
+
+        // Consume one character.
+        left_it++;
+        right_it++;
+    }
+}
+
+TEST(FormatHiliting, MetaTestConsumeHilites)
+{
+    using namespace DB;
+    std::stringstream ss;
+    // The order is different from the order in HILITES on purpose.
+    ss << IAST::hilite_keyword
+       << IAST::hilite_alias
+       << IAST::hilite_identifier
+       << IAST::hilite_none
+       << IAST::hilite_operator
+       << IAST::hilite_substitution
+       << IAST::hilite_function
+       << "test" << IAST::hilite_keyword;
+    std::string string = ss.str();
+    const char * it = string.c_str();
+    const char * expected_it = strchr(it, 't');
+    const char * last_hilite = consume_hilites(&it);
+    ASSERT_EQ(expected_it, it);
+    ASSERT_TRUE(last_hilite != nullptr);
+    ASSERT_EQ(IAST::hilite_function, last_hilite);
+}
+
+TEST(FormatHiliting, MetaTestRemoveHilites)
+{
+    using namespace DB;
+    std::stringstream ss;
+    ss << IAST::hilite_keyword
+       << "te" << IAST::hilite_alias << IAST::hilite_identifier
+       << "s" << IAST::hilite_none
+       << "t" << IAST::hilite_operator << IAST::hilite_substitution << IAST::hilite_function;
+    ASSERT_EQ("test", remove_hilites(ss.str()));
+}
+
+TEST(FormatHiliting, MetaTestAreEqualWithHilites)
+{
+    ASSERT_PRED2(are_equal_with_hilites, "", "");
+
+    for (const char * hilite : HILITES)
+    {
+        ASSERT_PRED2(are_equal_with_hilites, "", std::string_view(hilite));
+        ASSERT_PRED2(are_equal_with_hilites, std::string_view(hilite), "");
+    }
+
+    {
+        std::stringstream ss;
+        ss << DB::IAST::hilite_none << "select" << DB::IAST::hilite_none;
+        ASSERT_PRED2(are_equal_with_hilites, ss.str(), "select");
+    }
+
+    {
+        std::stringstream ss;
+        ss << DB::IAST::hilite_none << "\n " << "sel" << DB::IAST::hilite_none << "ect" << DB::IAST::hilite_none;
+        ASSERT_PRED2(are_equal_with_hilites, ss.str(), "\n select");
+    }
+
+    {
+        std::stringstream left;
+        left << DB::IAST::hilite_keyword << "keyword" << " long" << DB::IAST::hilite_none;
+        std::stringstream right;
+        right << DB::IAST::hilite_keyword << "keyword" << DB::IAST::hilite_none << " " << DB::IAST::hilite_keyword << "long";
+        ASSERT_PRED2(are_equal_with_hilites, left.str(), right.str());
+    }
+}
+
+void compare(const std::string & query, const std::stringstream & expected)
```

Comment:
Minor: ClickHouse provides typedef `String` which is a bit shorter than `std::string`. Alternatively use a `std::string_view` (which is conceptually a non-owning view to a string and can be passed as is, e.g. `void compare(std::string_view query, ...)`).

Reply:
Using String, as parseQuery requires a String, not std::string_view.